cpp/src/arrow/compute/kernels/take.cc

```diff
@@ -0,0 +1,221 @@
+  if (indices.null_count() == 0) {
+    return TakeImpl<OutOfBounds, AllValuesValid, true>(context, values, indices, builder);
+  }
+  return TakeImpl<OutOfBounds, AllValuesValid, false>(context, values, indices, builder);
+}
+
+template <int OutOfBounds, typename ValueArray, typename IndexArray, typename OutBuilder>
+Status TakeImpl(FunctionContext* context, const ValueArray& values,
+                const IndexArray& indices, OutBuilder* builder) {
+  if (values.null_count() == 0) {
+    return TakeImpl<OutOfBounds, true>(context, values, indices, builder);
+  }
+  return TakeImpl<OutOfBounds, false>(context, values, indices, builder);
+}
+
+template <typename ValueArray, typename IndexArray, typename OutBuilder>
+Status TakeImpl(FunctionContext* context, const ValueArray& values,
+                const IndexArray& indices, const TakeOptions& options,
+                OutBuilder* builder) {
+  switch (options.out_of_bounds) {
+    case TakeOptions::ERROR:
+      return TakeImpl<TakeOptions::ERROR>(context, values, indices, builder);
+    case TakeOptions::TONULL:
+      return TakeImpl<TakeOptions::TONULL>(context, values, indices, builder);
+    case TakeOptions::UNSAFE:
+      return TakeImpl<TakeOptions::UNSAFE>(context, values, indices, builder);
+    default:
+      ARROW_LOG(FATAL) << "how did we get here";
+      return Status::NotImplemented("how did we get here");
+  }
+}
+
+template <typename IndexType>
+struct UnpackValues {
+  using IndexArrayRef = const typename TypeTraits<IndexType>::ArrayType&;
+
+  template <typename ValueType>
+  Status Visit(const ValueType&) {
+    using ValueArrayRef = const typename TypeTraits<ValueType>::ArrayType&;
+    using OutBuilder = typename TypeTraits<ValueType>::BuilderType;
+    IndexArrayRef indices = static_cast<IndexArrayRef>(*params_.indices);
+    ValueArrayRef values = static_cast<ValueArrayRef>(*params_.values);
+    std::unique_ptr<ArrayBuilder> builder;
+    RETURN_NOT_OK(MakeBuilder(params_.context->memory_pool(), values.type(), &builder));
+    RETURN_NOT_OK(builder->Reserve(indices.length()));
+    RETURN_NOT_OK(TakeImpl(params_.context, values, indices, params_.options,
+                           static_cast<OutBuilder*>(builder.get())));
+    return builder->Finish(params_.out);
+  }
+
+  Status Visit(const NullType& t) {
+    auto indices_length = params_.indices->length();
+    if (params_.options.out_of_bounds == TakeOptions::ERROR && indices_length != 0) {
+      auto indices = static_cast<IndexArrayRef>(*params_.indices).raw_values();
+      auto max = *std::max_element(indices, indices + indices_length);
+      if (static_cast<int64_t>(max) > params_.values->length()) {
+        return Status::Invalid("out of bounds index");
+      }
+    }
+    params_.out->reset(new NullArray(indices_length));
+    return Status::OK();
+  }
+
+  Status Visit(const DictionaryType& t) {
+    auto dictionary_indices = params_.values->data()->Copy();
```

Comment:
This function requires more commenting.

```
template <typename ValueIndexArray>
std::shared_ptr<ValueIndexArray> ExtractNumericArrayFromDictionary(DictionaryArray &input) {
...
}

Status Visit(const DictionaryType& t) {
    TakeParameters params = params_;
    UnpackValues<IndexType> unpack = {params};

    // Creates a temporary numeric array on which the Take kernel operates
    unpack.params.values = ExtractNumericArrayFromDictionary(params_.values);

    RETURN_NOT_OK(VisitTypeInline(*t.index_type(), &unpack));

    // Casting back to the original array type
    *params_.out = std::make_shared<DictionaryArray>(...);
    return Status::OK()
}
```

Reply:
@fsaintjacques I have refactored this a bit, how does it look now?

Reply:
https://github.com/apache/arrow/pull/3880/files#diff-dbc961df861bbaa965256e5ad42236ceR168